docs_src/design/adr/0026-Common Configuration.md

```diff
@@ -0,0 +1,393 @@
+
+## Change Log
+- [pending](URL of PR TBD) (2022-11-29)
+
+## Referenced Use Case(s)
+- [Common Configuration UCR](URL of use case link TBD)
+
+## Context
+This ADR describes the architecture of  the new common configuration capability which impacts all services. Requirements for this new capability are describe in the above referenced UCR. This is deemed architecturally significant due to the cross cutting impacts.
+
+### Current Design
+
+The following flow chart the bootstrapping of each services' configuration in the current Levski release.
+
+![](common-config-images/EdgeX 2.x Configuration bootstrapping flowchart.png)
+
+## Proposed Design
+The configuration settings that are common to all services will be partitioned out into a separate common configuration source. This  common configuration source will be pushed into the Configuration Provider by a new service. When the Configuration Provider is not used this new service will provide the common configuration via a RESTful endpoint.
+
+During bootstrapping, each service will either load the common configuration from the Configuration Provider or via URI to some endpoint that provides the common configuration. This could be the new service's REST endpoint. Each service will have additional private configuration, which may overwrite and/or extend the common configuration.  
+
+An additional common configuration setting must be present to indicate all other common settings have been pushed to the Configuration Provider. This setting is stored last and the services must wait for this setting to be present prior to pulling the common settings.
+
+### Configuration Bootstrapping
+
+The following flow chart demonstrates the bootstrapping of each services' configuration with this new common configuration capability.
+
+![](common-config-images/EdgeX 3.x Configuration bootstrapping flowchart.png)
+
+### Secret Store Configuration
+
+As part of this design, the Secret Store configuration is being removed from the service configuration (common and private). This is so the Secret Provider can be instantiated prior to processing the service's configuration which may require the Secret Provider. The Secret Store configuration will now be a combination of default values and environment variable overrides.
+
+### Specifying the Common Configuration location
+
+If the `-cp/--configProvider` command line option is used the service will default to pulling the common configuration from a standard path in the Configuration Provider. i.e. `edgex/3.0/common/`
+
+The new `-cc/--commonConfig` command line option will be added for all services. This option will take the URI that specifies where the common configuration is pulled when not using the Configuration Provider. If no is URI specified with the option (ie. just `-cc`), the URI will default to the new configuration service's REST API. In addition a new environment override variable will be added which allows overriding this new command line option.

```

Comment:
The config provider value has the `type.URI` format, for example: `consul.http://localhost:8500` (from [docs](https://docs.edgexfoundry.org/2.3/microservices/configuration/ConfigurationAndRegistry/#seeding-configuration)). How about defining a new class of config providers for when a basic provider is used instead of Consul? The basic provider client needs to implement a subset of CP functionality and may need to support authentication (token based, basic auth, etc) as with Consul.

Reply:
For example, assuming the new config class is called `basic`, a file-based basic CP would be specified as `basic.file:///home/user/config` and expect to include an appropriate file structure beneath it (similar to that of Consul):
```
├── 3.0
    ├── common.json
    └── private
        ├── core-data.json
        ├── core-command.json
        └── ...
```

For an http-based CP, that translated to the following paths for a provider at `basic.http://example.com/edgex`:
- `http://example.com/edgex/common.json`
- `http://example.com/edgex/private/core-data.json`
- `http://example.com/edgex/private/core-command.json`
- `http://example.com/edgex/...`

Reply:
@farshidtz , Thinking about this, the basic file based CP implementation would use the existing file location option/defaults for the private settings and then the UIR for the common settings would be only info needed in the -cp value.
example: `basic.http://localhost:<port>/<path?`

Is this what you are suggesting?

Reply:
I recognise that the file-based example makes little sense because the private config can already be available at the default local path. My explanation was just to allow secure loading of configuration from an http-based provider which may include just the common stuff, or also private configs.

Thinking about this again, the `-cc/--commonConfig` flag is good if we limit the http-based access control to basic-auth for because the credentials can be passed in the URI (e.g. `-cc=https://user:pass@example.com/common.json`). So, no token based or other fancy authentication and client implementations, just `http.Get`.